Yoo

yooookm opened the pull request.

yooookm committed:
docs: 기능 명세서 작성

yooookm committed:
docs: 기능명세서 수정
역할을 작게 나눔
- 당첨번호 입력을 당첨번호 입력과 보너스 번호 입력으로 나눔
- 당첨 통계 계산을 로또 당첨 개수와 총 수익률 계산으로 나눔

yooookm committed:
test(LottoTest): 로또 유효성 검사
로또 번호에 대한 유효성 검사 테스트 코드 추가

yooookm committed:
feat(Lotto): 로또 유효성 검사 기능 추가
- 중복된 숫자 캐치
- 로또 번호가 1-45 안에 해당하는지 캐치

yooookm committed:
feat(ErrorConstants): 에러 메세지 상수 처리

yooookm committed:
test(LottoTest): 에러 메세지를 포함한 로또 유효성 처리 코드 수정

yooookm committed:
docs: 보너스 번호에 대한 기능 명세 수정
보너스 번호 유효성 검사 조건 추가

yooookm committed:
test(InputValidator): 구입 금액에 대한 테스트코드 추가
- 입력된 금액이 숫자인지 확인
- 정수 허용값 범위 내인지 확인
- 1000원 이상인지 확인
- 1000원 단위로 나뉘어지는지 확인

yooookm committed:
test(InputValidator): 당첨 번호 유효성 검사 테스트 코드 추가
- 당첨 번호가 쉼표로 구분되었는지 체크
- 당첨 번호가 숫자로 구성되었는지 체크
- 당첨번호가 6개로 구성되었는지 체크
- 당첨번호 입력에 쉼표가 중복되거나 입력값 앞, 뒤에 있는지 체크
- 당첨번호가 정수 허용값을 넘어섰는지 체크

yooookm committed:
docs: 당첨 번호 유효성 검사 조건 수정
- 중복값 확인 삭제
- 모든 번호가 정수 허용값 범위 내에 들어가는지 확인

yooookm committed:
test(InputValidator): 정상 입력을 확인하는 테스트 코드 추가
- 구입 금액 정상 입력 테스트 코드 추가
- 당첨 번호 정상 입력 테스트 코드 추가

yooookm committed:
test(InputValidator): 보너스 번호 유효성 검사 코드 추가
- 보너스 번호가 당첨 번호와 중복 되는지 확인
- 보너스 번호가 숫자인지 확인
- 보너스 번호가 1-45 범위인지 확인
- 올바른 입력일 때 예외를 던지지 않는지 확인

yooookm committed:
docs: 사용자 입력값 유효성 검사 조건 수정
- 구입 금액 정수 범위값 확인 조건 삭제
- 당첨 금액 정수 범위값 확인 조건 삭제

yooookm committed:
feat(InputValidator): 구입금액 입력 유효성 처리 기능 추가

yooookm committed:
feat(InputValidator): 구입금액 입력 유효성 처리 기능 추가

yooookm committed:
feat(InputValidator): 당첨번호 유효성 검사 기능 추가

yooookm committed:
feat(InputValidator): 보너스번호 유효성 검사 기능 추가

yooookm committed:
feat(LottoInputController): 사용자 입력 기능 추가
구입금액 입력 기능 추가
- IllegalArgumentException 오류가 발생했을 때, 오류 메세지 출력후 재입력
당첨번호 입력 기능 추가
 - IllegalArgumentException 오류가 발생했을 때, 오류 메세지 출력후 재입력
보너스 번호 입력 기능 추가
 - IllegalArgumentException 오류가 발생했을 때, 오류 메세지 출력후 재입력

yooookm committed:
feat(Lotto): 로또 랜덤 생성 기능 추가

yooookm committed:
feat(LottoManager): 사용자의 구입금액에 따른 로또 생성 기능

yooookm committed:
feat(LottoBank): Lotto 당첨순위 Enum 클래스 구현
- 1-6등  상수 선언
- 일치 개수와 보너스 일치 여부에 따른 순위 리턴

yooookm committed:
test(LottoManagerTest): 일치하는 번호개수에 따른 순위 계산

yooookm committed:
feat(LottoManager): 일치하는 번호개수에 따른 순위 계산 기눙 추가
- 입력받은 로또 번호와 당첨번호 비교
- 입력받은 로또 번호와 보너스번호 비교

yooookm committed:
refactor(LottoService): 로또 서비스 역할 분리
로또에 관련된 서비스 로직과 통계에 관련된 서비스 로직 분리
- LottoManager을 LottoService로 클래스명 변경
- LottoService 싱글톤 패턴 제거

yooookm committed:
test(LottoStatistics): 당첨 통계 테스트 코드 추가
- 등수별 당첨 횟수 계산 테스트 코드 추가
- 총 당첨금액 계산 테스트코드 추가
- 총 당첨금액을 이용한 수익률 계산 테스트코드 추가

yooookm committed:
feat(LottoStatistics): 각 등수별 당첨 횟수 계산 기능 추가

yooookm committed:
feat(LottoStatistics): 총 수익률 계산 기능 추가
- 총 당첨 금액을 이용해 수익률을 계산한다

yooookm committed:
feat(LottoResultPrinter): 당첨 통계 출력 기능 추가
- 당첨 통계 리스트와 총 수익률을 받아와 사용자에게 출력

yooookm committed:
refactor(LottoStatistics): 생성자 로직 추가
함수 순서 강제성 부여
- calculateWinsPerCategory 실행 후 calculateRateOfReturn 실행
- 당첨통계, 수익률 getter 함수 추가

yooookm committed:
refactor(LottoService): 싱글톤 패턴 적용
- 싱글턴 패턴 적용에 따른 테스트코드 수정
- 싱글톤 패턴 적용에 따른 LottoStatistics 생성자 수정

yooookm committed:
feat(Application): 전체 게임 진행 기능 추가
필요한 클래스 할당
사용자 입력값 받기
로또 생성 함수에 사용자 입력값 주입
로또 통계 계산 함수에 생성된 로또 주입
결과 프린트

yooookm committed:
feat(LottoResultPrinter): 출력 문자열 개행 추가

yooookm committed:
fix: Console 자원 수거

yooookm committed:
test(InputValidatorTest): 콘솔 자원 회수 코드 제거:

DAN-MU-ZI commented on src/main/java/lotto/Lotto.java:
여기서 매직넘버 처리를 추가하시면 더 깔끔할거 같아요.

DAN-MU-ZI commented on src/main/java/lotto/controller/InputValidator.java:
깔끔하게 작성하셨네요.

DAN-MU-ZI commented on src/main/java/lotto/domain/LottoRank.java:
enum의 활용법이 미숙했는데 어떻게 해야할지 알겠네요. 잘 하셨습니다.

DAN-MU-ZI commented on src/main/java/lotto/util/ErrorConstants.java:
error문에서 반복되는 부분이 있는데 좀더 줄일 수 있는 방법이 있을지 고민되네요.

yooookm replied:
[ERROR]만 매직 넘버로 쓰는 방법도 고민해봐야겠네요. 감사합니다!

DAN-MU-ZI reviewed (commented):
전반적으로 깔끔하게 작성하셔서 딱시 드릴 리뷰가 없네요.
고생하셨습니다.

MinseokGo commented on src/main/java/lotto/Application.java:
```diff
@@ -1,7 +1,24 @@
+        LottoInputController inputController = LottoInputController.getInstance();
+        LottoService lottoService = LottoService.getInstance();
+
+        int amount = inputController.getUserInputForAmount();
```
객체 값을 외부에서 관리하기에 부담스러워서 금액, 로또 등 값들을 캡슐화하니 객체 의존도가 너무 높아져서 고민입니다.

경미님처럼 값들을 외부에서 관리하면 코드 작성하기가 훨씬 수월해지겠지만 캡슐화도 한번 고민해보시면 좋을거 같아요!
(이건 저도 몰라서..)

yooookm replied:
객체 값을 캡슐화로 해야한다는 생각은 못했는데, 저도 한번 고려해봐야겠어요
확실히 외부에서 관리하면 위험할 것 같긴 하네요

MinseokGo commented on src/main/java/lotto/Lotto.java:
```diff
@@ -1,20 +1,52 @@
+    private List<Integer> numbers;
 
-    public Lotto(List<Integer> numbers) {
+    Lotto(List<Integer> numbers) {
```
default로 접근 지정을 하신 이유가 있을까요? 궁금합니다!

yooookm replied:
Lotto 객체를 외부에서 생성하는 것을 막고 싶은데, 테스트에서는 어쩔 수 없이 Lotto 객체를 생성해야 하다 보니 택한 방법입니다..ㅎㅎ
저번에는 protected로 선언했었는데, 이번에 조금 더 접근 범위를 좁혔습니다!
민석님은 테스트코드에서 private 함수를 테스트 하실 때 어떻게 하시나요?

MinseokGo replied:
저라면 객체 내부의 메서드가 값에 대한 상태성을 가지지 않는다면 public 접근 지정을 합니다.
객체 멤버 값에 대한 접근을 할 때에는 private 접근 지정을 해서 public 메서드에서 private 메서드를 간접 접근 하도록 구현하는데요.

이런 방식은 메서드마다 기능을 테스트해볼 수 없는 게 단점인 거 같습니다..

yooookm committed:
refactor(LottoGame): 불필요한 클래스 삭제

MinseokGo commented on src/main/java/lotto/Lotto.java:
```diff
@@ -1,20 +1,52 @@
 
-    private void validate(List<Integer> numbers) {
+    public static Lotto generate() {
+        List<Integer> numbers = new ArrayList<>(Randoms.pickUniqueNumbersInRange(1, 45, 6));
```
static 키워드를 사용하신 이유가 궁금합니다!

저같은 경우는 6이란 숫자는 로또의 정해진 사이즈이니 매직 넘버로 관리하는 게 편했습니다(1이나 45도 마찬가지였어요)

yooookm replied:
new 키워드로 Lotto를 생성하는 것을 막고, generate을 통해서만 로또가 생성될 수 있도록 만들기 위함입니다!
Lotto class 에서는 오직 랜덤 로또만 가지기 때문에 이 방법을 택했습니다.

MinseokGo commented on src/main/java/lotto/Lotto.java:
```diff
@@ -1,20 +1,52 @@
+        if (new HashSet<>(numbers).size() != 6) {
+            throw new IllegalArgumentException(DUPLICATE_NUMBER);
+        }
+        for (int number : numbers) {
```
stream을 사용해서 유효성 검증을 할 수도 있겠네요

yooookm replied:
훨씬 더 깔끔한 방법인 것 같네요 감사합니다!

MinseokGo commented on src/main/java/lotto/Lotto.java:
```diff
@@ -1,20 +1,52 @@
+
+    @Override
+    public String toString() {
+        return numbers.toString();
```
저는 이부분을 대괄호로 직접 하드 코딩 했는데 이런 방법도 있었네요! 배우고 갑니다.

MinseokGo commented on src/main/java/lotto/controller/InputValidator.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    private static void isCommaSeparated(String input) {
+        if (!input.matches("^(\\d+,){5}\\d+$")) {
```
정규 표현식도 어떤 표현식을 나타내는지 상수화 하면 더 읽기 좋은 코드가 될거 같아요

yooookm replied:
저번 주에도 피드백 받았던건데, 귀찮아서 나중에 해야지~ 하고 미루다 결국 잊어버렸네요.
저 표현식은 당첨 번호 포맷이 맞는지 확인하는 용도입니다. 저렇게 하면 ',' 중복이나 숫자 개수 검사를 안 해도 되더라고요!

MinseokGo commented on src/main/java/lotto/controller/LottoInputController.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.List;
+
+public class LottoInputController {
+    private static final LottoInputController inputController = new LottoInputController();
```
싱글톤 좋습니다👍👍

yooookm replied:
저번 주에 민석님 코드 보고 공부해서 적용해봤습니다 ㅎㅎ

MinseokGo commented on src/main/java/lotto/controller/LottoInputController.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public int getUserInputForAmount() {
+        while (true) {
```
while(true)는 경미님 코드를 읽는 다른 사람들이 이해하기 어렵지 않을까요?

MinseokGo commented on src/main/java/lotto/domain/LottoRank.java:
```diff
@@ -0,0 +1,46 @@
+        return bonusMatch;
+    }
+
+    public static LottoRank valueOf(int matchCount, boolean matchBonus) {
```
정적 팩토리 메서드 패턴을 이용하신 거 같네요 좋습니다👍👍

MinseokGo commented on src/main/java/lotto/domain/LottoService.java:
```diff
@@ -0,0 +1,37 @@
+    public LottoRank checkWinning(List<Integer> userNumbers, List<Integer> winningNumbers, int bonusNumber) {
+        int matchCount = (int) userNumbers.stream()
+                .filter(winningNumbers::contains)
+                .count();
```
stream을 사용하셔서 읽기 편했습니다👍👍

MinseokGo commented on src/main/java/lotto/domain/LottoStatistics.java:
```diff
@@ -0,0 +1,57 @@
+    private List<Integer> winningNumbers;
+    private int bonusNumber;
+    private LottoService lottoService;
+    private long totalRevenue = 0;
```
생성자를 사용하셨다면 생성자에서 하셔도 좋을거 같아요! 일관성이 더 좋아보이네요

MinseokGo commented on src/main/java/lotto/domain/LottoStatistics.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    private List<Integer> calculateWinsPerCategory() {
+        int[] wins = new int[LottoRank.values().length];
```
Map을 사용해보시는 건 어떨까요? 인덱스를 통해 당첨 카운트를 하시려는 것 같은데
Map<LottoRank, Integer>을 통해 가독성을 높이고 이후에 stream을 사용할 수 있다는 장점이 있을거 같아요!

yooookm replied:
그 방법은 생각 못했었는데, 한번 적용해봐야겠어요 감사합니다!

MinseokGo reviewed (commented):
이번주도 고생 많으셨습니다! 

저번주보다 더 편하게 코드를 읽을 수 있었습니다.

항상 코드를 작성하면서 고민되는 부분이 있는데 static과 final 키워드를 가장 고민하게 되는 것 같아요.

또한 캡슐화를 통해서 얻을 수 있는 장점도 많지만 코드 작성의 간결함을 포기해야하는 리스크가 너무 커서 고민의 굴레에 빠져있는 거 같습니다.

코드 리뷰 남겼으니 확인해보시고 편하게 의견주시면 좋을 거 같아요. 화이팅입니다!

MinseokGo commented on src/main/java/lotto/Application.java:
```diff
@@ -1,7 +1,24 @@
+        List<Integer> winningNumbers = inputController.getUserInputForWinningNumbers();
+        int bonusNumber = inputController.getUserInputForBonusNumber(winningNumbers);
+
+        LottoStatistics statistics = new LottoStatistics(lottos, winningNumbers, bonusNumber);
```
유저 로또, 당첨 로또 등을 관리하는 객체 같습니다.
위에서는 생성된 로또를 바로 출력하는데 이렇게 한다면 객체값의 무결성이 보장될 수 있을까요?
(로또 번호를 출력하고 로또 리스트가 변경된다면 무결성 위배가 될 수도 있을 거 같아요!)

yooookm replied:
final 처리를 고민해봐야겠네요.
생각해본 적 없는 부분이었는데, 짚어주셔서 감사합니다!

yooookm committed:
refactor(LottoStatistics): 생성자 로직 추가
- totalRevenue를 생성자 함수에서 초기화

yooookm committed:
refactor(Application): 객체 final 키워드 추가
- 객체 무결성 보장을 위한 final 키워드 추가

yooookm committed:
refactor(Lotto): 로또 범위 확인 방식 변경
- 기존 if문을 통한 검증에서 stream과 optional을 이용한 검증으로 변경

yooookm committed:
refactor(InputValidator): 정규 표현식 상수 처리
- 입력 포맷 확인을 위한 정규 표현식을 코드 가독성을 위해 상수 처리함

yooookm committed:
refactor(LottoStatistics): 당첨 확인 함수 기능 분리
객체지향 설계를 위한 당첨 확인 함수 기능 분리
- 당첨 확인과 총 상금 액수 계산 함수로 분리
- 함수 분리 위해 당첨 횟수 리스트 Map으로 변경
- 자료형 변경으로 따른 test 코드와 LottoResultPrinter.presentResults() 메서드 코드 변경

yooookm committed:
refactor(InputView,LottoInputController): 입력view 구분
- 기존 LottoInputController에 있던 메세지 출력 기능을 InputView로 분리

yooookm committed:
docs: 기능 명세서 수정
- 당첨 금액 계산과 총 수익률 계산 분리

yooookm committed:
refactor(LottoStatistics): 총 수익률 계산 로직 수정
- 총 수익률 계산 로직 안에서 당첨금액 계산 함수 실행

yooookm committed:
refactor(LottoInputController): 콘솔 지원 회수 코드 제거

yooookm committed:
refactor(LottStatistics): 당첨번호, 보너스 번호 무결성 보장
- 객체의 값이 바뀌는 것을 방지하기 위해 unmodifiableList로 할당함

yooookm committed:
docs: 기능 명세서 수정
- 공통 유효성 처리와 객체 유효성 처리의 명확한 구분

yooookm committed:
test: 파라미터화된 string 테스트 코드 테스트
